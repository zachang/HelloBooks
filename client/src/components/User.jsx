import React from 'react';
import PropTypes from 'react-proptypes';
import ReactDOM from 'react-dom';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Pagination } from 'react-materialize';
import { getBookAction, borrowBookAction } from '../actions/bookAction';
import BookCardUser from './book/BookCardUser.jsx';
import { decodeToken } from '../utils/helpers';
import { getCategoryAction } from '../actions/categoryAction';

/**
 * User class declaration
 * @class User
 * @extends {React.Component}
 */
export class User extends React.Component {
  /**
   * class constructor
   * @param {object} props
   */
  constructor(props) {
    super(props);
    this.state = {
      errors: null,
      categories: [],
      categoryId: '',
      pageCount: null,
      limit: 3,
      showToast: false,
    };
    this.borrowBook = this.borrowBook.bind(this);
    this.bookCategoryChange = this.bookCategoryChange.bind(this);
  }

  /**
   * @method componentWillMount
   * @return {void} void
   */
  componentWillMount() {
    this.props.getBookAction(this.state.limit, 0, this.state.categoryId);
    this.props.getCategoryAction();
  }

  /**
   * @method componentWillReceiveProps
   * @param {object} nextProps - nextProps
   * @return {object} nextProps
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.bookState.success === false) {
      this.setState({ errors: nextProps.bookState.errors });
    }

    if (nextProps.bookState.borrows === 'Borrow completed') {
      if (this.state.showToast) {
        Materialize.toast('Book borrowed, collect within 24hrs to prevent reversal!', 10000);
        this.setState({ showToast: false });
      }
    }

    if (nextProps.bookState.fails === 'You have not returned the previous book you borrowed') {
      if (this.state.showToast) {
        Materialize.toast('Please, return the previous book you borrowed', 5000);
        this.setState({ showToast: false });
      }
    } else if (nextProps.bookState.fails === 'All books have been borrowed') {
      if (this.state.showToast) {
        Materialize.toast('All copies have been borrowed, try again later!', 5000);
        this.setState({ showToast: false });
      }
    } else if (nextProps.bookState.fails === 'Process failed') {
      if (this.state.showToast) {
        Materialize.toast('Borrow failed, try again!', 5000);
        this.setState({ showToast: false });
      }
    }

    if (this.state.categories !== nextProps.categoryState.categories) {
      this.setState({ categories: nextProps.categoryState.categories });
    }
    if (this.state.pageCount !== nextProps.bookState.pageCount) {
      this.setState({ pageCount: nextProps.bookState.pageCount });
    }
  }

  /**
   * @method componentDidUpdate
   * @return {void} void
   */
  componentDidUpdate() {
    $('select').material_select();
    $('.tooltipped').tooltip({ delay: 50 });
    $(ReactDOM.findDOMNode(this.refs.categoryId)).on('change', this.bookCategoryChange.bind(this));
  }

  /**
   * @method bookCategoryChange
   * @return {void}
   * @param {object} event - event
   */
  bookCategoryChange(event) {
    this.setState({ categoryId: event.target.value });
    this.props.getBookAction(this.state.limit, 0, event.target.value);
  }

  /**
   * Handles book borrow
   * @method borrowBook
   * @param {string} token
   * @param {int} bookId
   * @return {void} void
   */
  borrowBook(token, bookId) {
    const userId = decodeToken(token).id;
    this.props.borrowBookAction(userId, bookId);
    this.setState({ showToast: true });
  }

  /**
   * Renders User component
   * @return {XML} JSX
   */
  render() {
    return (
      <div className='row'>
        <div className='section'>
          <h4 style={{ marginTop: '7%' }}>All Books</h4>
        </div>
        <div className='divider' style={{ marginTop: '-2%', marginBottom: '3%' }}></div>

        <div className='row'>
          <div className='input-field col s6 l4  m4'>
            <select
              name = 'categoryId'
              value = {this.state.categoryId}
              ref = 'categoryId'
              onChange= {this.bookCategoryChange}
            >
              <option value=''>Select Category</option>
              { this.state.categories.map((category, i) =>
                <option key={i} value={category.id}>{category.categoryName}</option>
              )}
            </select>
          </div>
        </div>

        <div className='row'>
          <div className='row'>
            { this.props.bookState.books.map((book, i) =>
              <BookCardUser
                key={i}
                book={book}
                borrowBook = {this.borrowBook}
              />
            )}
          </div>

          <div className='row'>
            {
              ((this.state.pageCount) ?
                <Pagination
                  items={this.state.pageCount}
                  onSelect={(page) => {
                    const offset = (page - 1) * this.state.limit;
                    this.props.getBookAction(this.state.limit, offset, this.state.categoryId);
                  }
                  } /> : '')
            }
          </div>

        </div>

      </div>
    );
  }
}

User.propTypes = {
  bookState: PropTypes.object.isRequired,
  categoryState: PropTypes.object.isRequired,
  getBookAction: PropTypes.func.isRequired,
  getCategoryAction: PropTypes.func.isRequired,
  borrowBookAction: PropTypes.func.isRequired
};
const mapStateToProps = state => ({
  bookState: state.bookReducer,
  categoryState: state.categoryReducer
});
const mapDispatchToProps = dispatch =>
  bindActionCreators({ getBookAction, getCategoryAction, borrowBookAction }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(User);