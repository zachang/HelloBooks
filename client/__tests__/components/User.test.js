import React from 'react';
import toJson from 'enzyme-to-json';
import sinon from 'sinon';
import { shallow, mount } from 'enzyme';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import storeConfiguration from '../../src/store/store';
import { User } from '../../src/components/User.jsx';

const event = {
  preventDefault: jest.fn(),
  target: {
    value: 'mmmmmm'
  }
};

describe('<User/>', () => {
  // create a spy function for getBookAction
  const getBookAction = sinon.spy();
  // create a spy function for getCategoryAction
  const getCategoryAction = sinon.spy();
  // create a spy function for clearGoogleSigninAction
  const clearGoogleSigninAction = sinon.spy();
  // create a spy function for borrowBookAction
  const borrowBookAction = sinon.spy();
  // spy on componentDidMount of User Page
  sinon.spy(User.prototype, 'componentDidMount');
  // spy on bookCategoryChange of User Page
  sinon.spy(User.prototype, 'bookCategoryChange');
  // spy on borrowBook of User Page
  sinon.spy(User.prototype, 'borrowBook');
  // spy on readBook of User Page
  sinon.spy(User.prototype, 'readBook');
  // spy on componentWillReceiveProps of User Page
  sinon.spy(User.prototype, 'componentWillReceiveProps');

  const props = {
    bookState: {
      success: null,
      errors: null,
      books: ['nnnnn'],
      pageCount: null,
      borrows: null,
      fails: null
    },
    categoryState: {
      success: false,
      errors: null,
      categories: []
    },
    signinState: {
      success: false,
      googleSigned: null,
      fails: null,
    },
    categoryId: '',
    getBookAction,
    getCategoryAction,
    clearGoogleSigninAction,
    borrowBookAction,
    pageCount: null,
    showToast: false,
  };

  const middleware = [thunk];
  const mockStore = configureMockStore(middleware);
  let store;
  let shallowComponent;
  beforeEach(() => {
    store = mockStore(storeConfiguration);
    shallowComponent = shallow(<User {...props} />);
    shallowComponent.setState({
      pageCount: 1,
      categories: ['mmm', 'mmmm'],
      showToast: false,
    });
  });

  it('renders <User /> component', () => {
    const tree = toJson(shallowComponent);
    expect(tree).toMatchSnapshot();
  });

  it('calls componentDidMount', () => {
    shallowComponent.instance().componentDidMount(props);
    expect(User.prototype.componentDidMount.called).toEqual(true);
  });

  it('should check that bookCategoryChange is called', () => {
    shallowComponent.instance().bookCategoryChange(event);
    expect(User.prototype.bookCategoryChange.called).toEqual(true);
  });

  it('should check that readBook is called', () => {
    shallowComponent.instance().readBook();
    expect(User.prototype.readBook.called).toEqual(true);
  });

  it('should check that borrowBook is called', () => {
    shallowComponent.instance().borrowBook(token, 1);
    expect(User.prototype.borrowBook.called).toEqual(true);
  });

  it('calls componentWillReceiveProps when books are retrieved', () => {
    props.bookState.success = false;
    shallowComponent.setState({
      errors: props.bookState.errors
    });
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(User.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it('calls componentWillReceiveProps when borrowing a book', () => {
    props.bookState.borrows = 'Borrow completed';
    shallowComponent.setState({
      showToast: true
    });
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(User.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it('calls componentWillReceiveProps when book not returned', () => {
    props.bookState.fails = 'You have not returned ' +
      'the previous book you borrowed';
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(User.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it('calls componentWillReceiveProps when book not available', () => {
    props.bookState.fails = 'All books have been borrowed';
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(User.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it('calls componentWillReceiveProps when borrow fails', () => {
    props.bookState.fails = 'Process failed';
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(User.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it('calls componentWillReceiveProps when gmail login', () => {
    props.signinState.googleSigned = 'Gmail login successful';
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(User.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it('calls componentWillReceiveProps when paginating', () => {
    props.pageCount = 1;
    props.bookState.pageCount = 2;
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(User.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it(`should check if the 'BookCardUser' is defined`, () => {
    expect(shallowComponent.find('BookCardUser').length).toBe(1);
  });

  it(`should check if the 'BookCardUser' is defined`, () => {
    expect(shallowComponent.find('BookCardUser').length).toBe(1);
  });

  it(`should check if the 'Pagination' is defined`, () => {
    expect(shallowComponent.find('Pagination').length).toBe(1);
    shallowComponent.find('Pagination').simulate('select');
  });
});
