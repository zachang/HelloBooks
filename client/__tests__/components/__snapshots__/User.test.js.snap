// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<User/> renders <User /> component 1`] = `
<div
  className="row"
>
  <div
    className="section"
  >
    <h4
      style={
        Object {
          "marginTop": "7%",
        }
      }
    >
      All Books
    </h4>
  </div>
  <div
    className="divider"
    style={
      Object {
        "marginBottom": "3%",
        "marginTop": "-2%",
      }
    }
  />
  <div
    className="row"
  >
    <div
      className="input-field col s6 l4  m4"
    >
      <select
        name="categoryId"
        onChange={[Function]}
        value=""
      >
        <option
          value=""
        >
          Select Category
        </option>
        <option
          key="0"
        />
        <option
          key="1"
        />
      </select>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="row"
    >
      <BookCardUser
        book="nnnnn"
        borrowBook={[Function]}
        key="0"
        readBook={[Function]}
      />
    </div>
    <div
      className="row"
    >
      <Pagination
        activePage={1}
        items={1}
        onSelect={[Function]}
      />
    </div>
  </div>
</div>
`;
